cmake/ImportBSON.cmake

```diff
@@ -0,0 +1,170 @@
+#[[
+   This file defines, exports, and installs two INTERFACE targets: '_mongocrypt::libbson_for_static'
+   and '_mongocrypt::libbson_for_shared', that are used to link libbson correctly for the build
+   configuration of libmongocrypt. At find_package() time, we can resolve these interface targets
+   to link to the appropriate libbson based on the build configurations of libmongocrypt.
+
+   mongo::mongocrypt must link to _mongocrypt::libbson_for_shared, and mongo::mongocrypt_static must
+   link to _mongocrypt::libbson_for_static.
+
+   At configure+build time, these target will create BUILD_INTERFACE-only usage requirements
+   appropriate for libmongocrypt to build against a libbson. Once these targets are installed,
+   they retain no usage requirements defined here.
+
+   Instead, the installed version of these targets will be manipulated in mongocrypt-config.cmake
+   based on user settings and build configuration options of the installed libmongocrypt in order
+   to ensure that users have satisfied the linking requirements of libmongocrypt.
+   Refer to mongocrypt-config.cmake for more information
+
+   This file calls add_subdirectory(EXCLUDE_FROM_ALL) on a mongo-c-driver project directory. This
+   will expose libbson targets that we can link and use for the libmongocrypt build.
+
+   The boolean option USE_SHARED_LIBBSON controls the behavior of libbson_for_shared:
+
+   If USE_SHARED_LIBBSON=FALSE:
+
+   - libbson_for_shared will transitively link the static libbson from the MONGOCRYPT_MONGOC_DIR.
+   - The result is that mongo::mongocrypt (which is a SHARED library) will have the translation
+     units of libbson directly embedded into the resulting binary.
+   - The symbols from libbson that are merged into mongo::mongocrypt will be supressed using
+     linker scripts such that consumers of mongo::mongocrypt will not see the libbson symbols that
+     were statically linked into the shared library. This allows consumers to link against a
+     completely independent libbson without interfering with the libbson symbols that were merged
+     into mongo::mongocrypt
+   - The installed libbson_for_shared will have no usage requirements.
+
+   If USE_SHARED_LIBBSON=TRUE:
+
+   - libbson_for_shared will transitively use the shared libbson library from
+     the MONGOCRYPT_MONGOC_DIR.
+   - mongo::mongocrypt will be built with a dynamic link requirement on a libbson dynamic
+     library, which must be resolved at runtime by consumers. The translation units from the
+     MONGOCRYPT_MONGOC_DIR *will not* be included in the mongo::mongocrypt library.
+   - The installed libbson_for_shared will dynamically link to a libbson on the user's system by
+     using a find_library() call.
+
+   In both of the above cases, libbson_for_static will require that the final consumer
+   provide their own definitions of the libbson symbols, regardless of the value
+   of USE_SHARED_LIBBSON.
```

Comment:
> @kevinAlbs: To confirm, is it is no longer possible to use a system install (e.g. link to /usr/local/lib/libbson.so)?

It is, but also isn't. In the case of `USE_SHARED_LIBBSON`, generated files will contain a dynamic-link requirement to a libbson dynamic library, referenced by a portable name (i.e. ELF uses the SONAME, DLL uses the filename, etc.). It is up to the runtime linker to decide what file will satisfy the reference and resolve the external symbols. The main purpose of the "linking" the dynamic libmongocrypt library to a libbson dynamic library is to embed this dynamic-link reference and ensure that the referrenced symbols actually exist in the dynamic library, but the dynamic link does not pin that dynamic library reference nor the resolved symbols to any particular instance of the dynamic library. In the current revision, the `USE_SHARED_LIBBSON` will compile the libbson from the subproject and then link against it. This will generate the appropriate runtime linking rules that a system's libbson will also be able to satisfy the dynamic link. Because libbson is very stable, and because libmongocrypt here links to libbson 1.17, any newer libbson-1.0 will satisfy the link, including a system's libbson.

---

> @rcsanchez97: It looks like there needs to be a way to tell the build "use the system libbson and don't go out to the network looking for it".

Is it required that the actual build-time `ld` invocation _specifically use_ the `libbson` installed in `/usr/lib`, or is the requirement that the generated runtime linker request be _satisfyable_ by the `libbson` in `/usr/lib`?

- If the former:
  - The `ImportBSON.cmake` can be tweaked to instead define `_mongocrypt-libbson_for_{static,shared}` to link against the `mongo::libbson` targets resolved via a `find_package()` call, as they were before. This would still require the intermediate build step to install a partial `mongo-c-driver` that omits the libmongocrypt dependency.
- If the latter:
  - The generated source-dist can include the fetch snapshot of `mongo-c-driver` and `ImportBSON.cmake` can `add_subdirectory()` that directory without invoking `FetchBSON`. This design would alleviate the requirement for the separate `pkg.mongo-c-driver.no-libmongocrypt` step, and the resulting binaries would still be satisfied by the dynamic `libbson` from the `libbson-1.0` package.

Reply:
In the Debian package and RPM builds (the official builds, not the packaging builds we run in Evergreen to create packages for the PPA), the process involves first installing the build-time dependencies (including libbson from a package which can be installed via apt or dnf, as applicable to the platform), then building libmongocrypt against that specific libbson.  In theory, I can't think of a reason why the build must be limited to linking to the specific libbson installed in `/usr/lib` by the build process.  However, in practice the limitation that the build complete without accessing remote network resources means that the only available libbson will be in `/usr/lib` when building Debian packages and RPMs.

The Debian and Fedora policies would not allow for the solution you are proposing.  I think the simplest way to state the requirement is "there is must be a way to build libmongocrypt with only system components and the components included in the libmongocrypt source tree itself (i.e., no remote network access from the build)."  Note that the C++ driver has a similar situation with the polyfill.  There is one build configuration which downloads MNMLSTC (and which is therefore unsuitable for Debian package and RPM builds), and another that allows building the C++ driver against and already installed Boost.

I have no objection to making the experience simpler for libmongocrypt users, but we will need a patch that allows us to meet the requirements of distribution packaging policies.

Reply:
I've added a special condition in `ImportBSON` that can be used to override the behavior for use by package maintainers.